R/file_cache.R

```diff
@@ -0,0 +1,542 @@
+    },
+
+    #' @description Return the number of items currently in the cache.
+    size = function() {
+      self$is_destroyed(throw = TRUE)
+      length(dir(private$dir))
+    },
+
+    #' @description Clears all objects in the cache, and removes the cache
+    #'   directory from disk.
+    destroy = function() {
+      if (self$is_destroyed()) {
+        return(invisible(self))
+      }
+
+      private$log(paste0("destroy: Removing ", private$dir))
+      # First create a sentinel file so that other processes sharing this
+      # cache know that the cache is to be destroyed. This is needed because
+      # the recursive unlink is not atomic: another process can add a file to
+      # the directory after unlink starts removing files but before it removes
+      # the directory, and when that happens, the directory removal will fail.
+      file.create(file.path(private$dir, "._destroyed__"))
+      # Remove all the cache files. This will not remove the sentinel file.
+      file.remove(dir(private$dir, full.names = TRUE))
+      # Next remove dir recursively, including sentinel file.
+      unlink(private$dir, recursive = TRUE)
+      private$destroyed <- TRUE
+      invisible(self)
+    },
+
+    #' @description Reports whether the cache has been destroyed.
+    #' @param throw Should this function throw an error if the cache has been
+    #'   destroyed?
+    is_destroyed = function(throw = FALSE) {
+      if (!dir.exists(private$dir) ||
+          file.exists(file.path(private$dir, "._destroyed__")))
+      {
+        # It's possible for another process to destroy a shared cache directory
+        private$destroyed <- TRUE
+      }
+
+      if (throw) {
+        if (private$destroyed) {
+          stop("Attempted to use cache which has been destroyed:\n  ", private$dir)
+        }
+
+      } else {
+        private$destroyed
+      }
+    },
+
+    #' @description A finalizer for the cache.
+    finalize = function() {
+      if (private$destroy_on_finalize) {
+        self$destroy()
+      }
+    }
+  ),
+
+  private = list(
+    dir = NULL,
+    max_age = NULL,
+    max_size = NULL,
+    max_n = NULL,
+    evict = NULL,
+    destroy_on_finalize = NULL,
+    destroyed = FALSE,
+    logfile = NULL,
+
+    prune_throttle_counter = 0,
+    prune_last_time = NULL,
+
+    filename_full_path = function(filename) {
+      file.path(private$dir, filename)
+    },
+
+    # A wrapper for prune() that throttles it, because prune() can be
+    # expensive due to filesystem operations. This function will prune only
+    # once every 20 times it is called, or if it has been more than 5 seconds
+    # since the last time the cache was actually pruned, whichever is first.
+    # In the future, the behavior may be customizable.
+    prune_throttled = function() {
+      # Count the number of times prune() has been called.
+      private$prune_throttle_counter <- private$prune_throttle_counter + 1
+
+      if (private$prune_throttle_counter > 20 ||
```

Comment:
Why 20 calls?

If they are misses, we could assume it would be 0.5 seconds to calculate. So 20 calls would take 5 seconds.  

If they are small or cache hits, then we can easily wait the 5 minutes.  

I put a vote towards removing the throttle counter.

Reply:
`prune_throttled()` is called by `set_file()` and `set_file()` (but not the `get` methods). The set methods don't check the cache; there are no cache hits involved.

These set methods could be called dozens, or hundreds of times within 5 seconds, and if no pruning occurs, it could cause the cache to balloon well past the intended size. The number 20 was chosen as a compromise to help prevent going way past that limit.